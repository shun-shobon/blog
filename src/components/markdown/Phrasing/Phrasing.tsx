import Image from "next/image";
import { type PhrasingContent } from "mdast";

export type PhrasingProps = {
  content: PhrasingContent;
};

const Phrasing = ({ content }: PhrasingProps): JSX.Element => {
  const f = (content: PhrasingContent, index: number) => (
    <Phrasing content={content} key={index} />
  );

  switch (content.type) {
    case "text":
      return <>{content.value}</>;
    case "emphasis":
      return <em>{content.children.map(f)}</em>;
    case "strong":
      return <strong>{content.children.map(f)}</strong>;
    case "delete":
      return <del>{content.children.map(f)}</del>;
    case "html":
      return <div dangerouslySetInnerHTML={{ __html: content.value }} />;
    case "inlineCode":
      return <code>{content.value}</code>;
    case "break":
      return <br />;
    case "image":
      return <Image src={content.url} alt={content.alt ?? undefined} />;
    case "link":
      return <a href={content.url}>{content.children.map(f)}</a>;
    case "imageReference":
    case "footnote":
    case "footnoteReference":
    case "linkReference":
      throw new Error("Not implemented");
  }
};

export default Phrasing;
