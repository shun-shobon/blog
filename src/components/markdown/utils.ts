import type { FootnoteDefinition, FootnoteReference } from "mdast";

export function getFootnoteIndex(
  footnoteReference: FootnoteReference,
  footnoteDefinitions: FootnoteDefinition[],
): number {
  const index = footnoteDefinitions.findIndex((footnoteDefinition) => {
    return footnoteDefinition.identifier === footnoteReference.identifier;
  });
  if (index === -1) {
    throw new Error("Footnote definition not found");
  }

  return index;
}

export function getFootnoteRefId(index: number): string {
  return `fnref-${index}`;
}

export function getFootnoteDefId(index: number): string {
  return `fn-${index}`;
}
