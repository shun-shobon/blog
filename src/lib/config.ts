export type Config = {
  isProduction: boolean;
  articlesPath: string;
};

export function getConfig(): Config {
  const isProduction = process.env.NODE_ENV === "production";
  const articlesPath = isProduction ? "articles" : "test/articles";

  return {
    isProduction,
    articlesPath,
  };
}
