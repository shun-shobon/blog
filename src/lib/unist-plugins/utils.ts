import type {
  Definition,
  Heading,
  ImageReference,
  Link,
  LinkReference,
  List,
  ListItem,
  Paragraph,
  Parent,
  Text,
  YAML,
} from "mdast";
import type { Node } from "unist";

import type { Section } from "./remark-section";

export function findNodeAfter<T extends Node>(
  tree: Parent,
  after: number,
  predicate: (node: Node) => node is T,
): [T, number] | [undefined, undefined] {
  const idx = tree.children.slice(after).findIndex(predicate);
  if (idx === -1) return [undefined, undefined];

  const index = idx + after;
  return [tree.children[index] as T, index];
}

export function isParent(node?: Node | null): node is Parent {
  return node != null && Object.hasOwn(node, "children");
}

export function isYAML(node?: Node | null): node is YAML {
  return node != null && node.type === "yaml";
}

export function isSection(node?: Node | null): node is Section {
  return node != null && node.type === "section";
}

export function isHeading(node?: Node | null): node is Heading {
  return node != null && node.type === "heading";
}

export function isHeadingOfDepth(depth: number) {
  return (node?: Node | null): node is Heading => {
    return isHeading(node) && node.depth === depth;
  };
}

export function isParagraph(node?: Node | null): node is Paragraph {
  return node != null && node.type === "paragraph";
}

export function isList(node?: Node | null): node is List {
  return node != null && node.type === "list";
}

export function isListItem(node?: Node | null): node is ListItem {
  return node != null && node.type === "listItem";
}

export function isText(node?: Node | null): node is Text {
  return node != null && node.type === "text";
}

export function isLink(node?: Node | null): node is Link {
  return node != null && node.type === "link";
}

export function isLinkReference(node?: Node | null): node is LinkReference {
  return node != null && node.type === "linkReference";
}

export function isImageReference(node?: Node | null): node is ImageReference {
  return node != null && node.type === "imageReference";
}

export function isDefinition(node?: Node | null): node is Definition {
  return node != null && node.type === "definition";
}
