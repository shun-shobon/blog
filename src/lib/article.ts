import type { ArticleSummaries, ArticleSummary } from "@/lib/markdown";
import { ARTICLE_SUMMARIES_FILE_NAME } from "@/lib/markdown";

import type { Article } from "./plugins";

export async function getAllArticleSummaries(): Promise<ArticleSummary[]> {
  return (
    (await import(`data/${ARTICLE_SUMMARIES_FILE_NAME}`)) as ArticleSummaries
  ).summaries;
}

export async function getArticleSummariesByTag(
  tag: string,
): Promise<ArticleSummary[]> {
  const summaries = await getAllArticleSummaries();

  return summaries.filter((summary) => summary.tags.includes(tag));
}

export async function getAllTags(): Promise<string[]> {
  const summaries = await getAllArticleSummaries();

  const tags = new Set<string>();
  summaries.forEach((summary) => summary.tags.forEach((tag) => tags.add(tag)));

  return Array.from(tags);
}

export async function getArticle(slug: string): Promise<Article> {
  return (await import(`data/${slug}.json`)) as Article;
}
