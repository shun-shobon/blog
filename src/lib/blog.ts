import * as path from "node:path";

import glob from "fast-glob";

import { getConfig } from "./config";

async function findArticles() {
  const { articlesPath } = getConfig();
  const articles = await glob(path.join(articlesPath, "*/main.md"));
}
