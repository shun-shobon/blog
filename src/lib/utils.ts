import { ARTICLE_REPOSITORY_EDIT_PATH, ORIGIN } from "@/config";

export class UnreachableError extends Error {
  constructor() {
    super();

    this.name = "UnreachableError";
  }
}

export function joinPath(...paths: string[]): string {
  return paths.join("/").replace(/\/+/g, "/").replace(/\.\//, "");
}

export function getUrl(...paths: string[]): URL {
  return new URL(joinPath(...paths), ORIGIN);
}

export function getArticleUrl(slug: string): URL {
  return getUrl("/articles", slug);
}

export function getTagUrl(name: string): URL {
  return getUrl("/tags", name);
}

export function getEditArticleUrl(slug: string): URL {
  return new URL(
    joinPath(ARTICLE_REPOSITORY_EDIT_PATH, slug, "README.md"),
    "https://github.com",
  );
}

export function bufferToHex(buffer: ArrayBuffer): string {
  return Array.from(new Uint8Array(buffer))
    .map((byte) => byte.toString(16).padStart(2, "0"))
    .join("");
}
