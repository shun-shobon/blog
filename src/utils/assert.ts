import * as E from "fp-ts/lib/Either";

export function assertRight<A>(
  e: E.Either<unknown, A>,
): asserts e is E.Right<A> {
  if (E.isLeft(e)) {
    throw new Error(`Expected Right, got Left: ${e.left}`);
  }
}
