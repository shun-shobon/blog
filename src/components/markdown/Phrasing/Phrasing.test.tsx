/** @jest-environment jsdom */

import { render } from "@testing-library/react";

import Phrasing from "./Phrasing";

import { assertRight } from "../../../utils/assert";
import { getArticle } from "../../../lib/blog";
import { type Paragraph } from "mdast";

describe("Phrasing", () => {
  it("renders", async () => {
    const article = await getArticle(
      "tests/articles",
      "2021-12-05/test1/main.md",
    );
    assertRight(article);
    const contents = (article.right.contents.children[11] as Paragraph)
      .children;

    const component = render(
      <>
        {contents.map((content, index) => (
          <Phrasing content={content} key={index} />
        ))}
      </>,
    );

    expect(component.getByText("Bold")).toBeInTheDocument();
    expect(component.getByText("Italic")).toBeInTheDocument();
    expect(component.getByText("Bold Italic")).toBeInTheDocument();
    expect(component.getByText("Inline Code")).toBeInTheDocument();
    expect(component.getByText("Strikethrough")).toBeInTheDocument();

    expect(component.asFragment()).toMatchSnapshot();
  });
});
