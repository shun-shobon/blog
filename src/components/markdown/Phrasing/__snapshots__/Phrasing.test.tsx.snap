// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phrasing renders 1`] = `
<DocumentFragment>
  <strong>
    Bold
  </strong>
   
  <em>
    Italic
  </em>
   
  <em>
    <strong>
      Bold Italic
    </strong>
  </em>
   
  <code>
    Inline Code
  </code>
   
  <del>
    Strikethrough
  </del>
</DocumentFragment>
`;
